### PR TITLE
fix: render orphan data points on lines and areas

### LINKED ISSUE #63: Display orphan data points on line/area series
**Is your feature request related to a problem? Please describe.**
Some series may have orphan data points, points that are between two null datums like:
```
[
   ...,
   {x: 1, y: null},
   {x: 2, y: 10},
   {x: 3, y: null},
   ...,
]
```
When a series has orphan data points like the example above, we need to find a way to display the single point value.

Fitting functions (#55) can fill the gaps with specific values (like adding zeros or an average value), but the developer or user can decide to keep `null` values.

**Describe the solution you'd like**
If the style of the line/area already shows single circles as data points we don't need to add anything to that rendering.

If the style configuration hides the data points we need to force displaying these points when orphans.

We also have to configure the line/area path generation `defined` function to correctly specify when a line/area is not defined: https://github.com/d3/d3-shape#line_defined https://github.com/d3/d3-shape#area_defined

**Kibana Cross Issues**
Add any [Kibana](https://github.com/elastic/kibana) related issues/feature requests here.

**Checklist**

- [x] this request is checked against already exist requests
- [ ] every related Kibana issue is listed under `Kibana Cross Issues` list
- [ ] `kibana cross issue` tag is associated to the issue if any kibana cross issue is present

### LINKED ISSUE #783: Display orphan data points on line/area charts also if the point style is invisible
I heard from Lens team this is how Elastic chart displays, so opening this issue here 
image

When in Lens and with sparse data the dops/points  on the chart shows empty even though once her on them they have value
The behavior doesn't make sense

### PR DESCRIPTION
**Pull request by markov00**
## Summary

This PR adds the ability to render orhpan data points displayed on a line or area chart.
An orphan data point is a datapoint that doesn't have a next or a previous data point to connect to.
The possible cases are:
- the first data-point on a series with a null, undefined consecutive datapoint
- the last data-point of a series with a null, undefined previous datapoint
- a data-point of a series with both previous and next points null.

If the series style is configured to show the point, it will not change the current rendering behaviour if, instead, the series style hides the point (`point:{visible: false}`) on the rendering phase we filter and render only the ophan data points.

Screenshot 2020-11-11 at 13 19 01
 

fix #783 
close #63 

### Checklist

- [x] Proper documentation or storybook story was added for features that require explanation or tutorials
- [x] Unit tests were updated or added to match the most common scenarios

### CODE REVIEW

**Review by monfera — COMMENTED** [review]
For aesthetic reasons and to reinforce the meaning of the marker circles, as well as emphasize the section starts and ends, maybe the section starts/ends could also have the marker.

File: src/chart_types/xy_chart/rendering/points.ts [hunk 1]
```diff
@@ -228,3 +235,26 @@ export function getRadiusFn(
     return circleRadius ? Math.sqrt(circleRadius + baseMagicNumber) + lineWidth : lineWidth;
   };
 }
+
+function yAccessorForOrphanCheck(datum: DataSeriesDatum): number | null {
+  return datum.filled?.y1 ? null : datum.y1;
+}
+
+function isOrphanData(
```
Comment (nickofthyme) [inline src/chart_types/xy_chart/rendering/points.ts:243]: maybe a little more specific?

```suggestion
function isOrphanDataPoint(
```

**Review by nickofthyme — APPROVED** [review]
LGTM, just a few minor suggestions. 👍

File: src/chart_types/xy_chart/rendering/points.ts [hunk 2]
```diff
@@ -59,8 +61,10 @@ export function renderPoints(
   const y0Fn = getY0ScaledValueOrThrow(yScale);
   const yDefined = isYValueDefined(yScale, xScale);
 
-  const pointGeometries = dataSeries.data.reduce((acc, datum) => {
+  const pointGeometries = dataSeries.data.reduce((acc, datum, dataIndex) => {
     const { x: xValue, mark } = datum;
+    const prev = dataSeries.data[dataIndex - 1];
+    const next = dataSeries.data[dataIndex + 1];
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/xy_chart/rendering/points.ts:67]: Yes

**Review by monfera — APPROVED** [review]
LGTM, achieves the objective, and the section start/end markers are beyond the scope of this PR

File: src/chart_types/xy_chart/rendering/utils.ts [hunk 3]
```diff
@@ -39,16 +39,15 @@ export interface MarkSizeOptions {
  * Passing a filled key (x, y1, y0) it will return that value or the filled one
  * @internal
  */
-export const getYDatumValue = (
+export const getYDatumValue = (valueName: keyof Omit<FilledValues, 'x'> = 'y1') => (
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/xy_chart/rendering/utils.ts:42]: actually is not returning a function, but a value

File: src/chart_types/xy_chart/rendering/points.ts (hunk 1 above)
Comment (markov00) [reply src/chart_types/xy_chart/rendering/points.ts]: done https://github.com/elastic/elastic-charts/pull/900/commits/b94d979f545839e78bbcd2eab59a421346ccea1c

File: src/chart_types/xy_chart/rendering/utils.ts [hunk 4]
```diff
@@ -170,14 +169,17 @@ const DEFAULT_ZERO_BASELINE = 0;
...
+  getValueAccessor: (datum: DataSeriesDatum) => number | null,
+) => boolean;
+
+/** @internal */
+export function isYValueDefined(yScale: Scale, xScale: Scale): YDefinedFn {
   const isLogScale = isLogarithmicScale(yScale);
   const domainPolarity = getDomainPolarity(yScale.domain);
-  return (datum, valueName = 'y1') => {
-    const yValue = getYDatumValue(datum, valueName);
+  return (datum, getValueAccessor: (datum: DataSeriesDatum) => number | null) => {
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/xy_chart/rendering/utils.ts]: done https://github.com/elastic/elastic-charts/pull/900/commits/b94d979f545839e78bbcd2eab59a421346ccea1c

File: src/chart_types/xy_chart/rendering/points.ts [hunk 5]
```diff
@@ -228,3 +235,26 @@ export function getRadiusFn(
...
+  return datum.filled?.y1 ? null : datum.y1;
+}
+
+function isOrphanData(
+  prev: DataSeriesDatum | undefined,
+  next: DataSeriesDatum | undefined,
+  index: number,
+  length: number,
+  yDefined: YDefinedFn,
+) {
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/xy_chart/rendering/points.ts]: done https://github.com/elastic/elastic-charts/pull/900/commits/b94d979f545839e78bbcd2eab59a421346ccea1c

[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/xy_chart/rendering/points.ts]: done https://github.com/elastic/elastic-charts/pull/900/commits/b94d979f545839e78bbcd2eab59a421346ccea1c

File: src/chart_types/xy_chart/rendering/utils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/xy_chart/rendering/utils.ts]: done https://github.com/elastic/elastic-charts/pull/900/commits/b94d979f545839e78bbcd2eab59a421346ccea1c

[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/xy_chart/rendering/utils.ts]: I've also renamed other functions that return functions with the same strategy

File: src/chart_types/xy_chart/renderer/canvas/lines.ts [hunk 6]
```diff
@@ -46,28 +46,27 @@ export function renderLines(ctx: CanvasRenderingContext2D, props: LineGeometries
...
+        panel,
+        rotation,
+        renderingArea,
+        (ctx) => {
+          const geometryStyle = getGeometryStateStyle(line.seriesIdentifier, sharedStyle, highlightedLegendItem);
+          renderPoints(ctx, visiblePoints, line.seriesPointStyle, geometryStyle);
+        },
+        // TODO: add padding over clipping
+        { area: clippings, shouldClip: line.points[0]?.value.mark !== null },
+      );
```
[the remark this answers is not in the document]
Comment (markov00) [reply src/chart_types/xy_chart/renderer/canvas/lines.ts]: I've added that check here https://github.com/elastic/elastic-charts/pull/900/commits/b94d979f545839e78bbcd2eab59a421346ccea1c
the performance gain is probably very small

**Review by nickofthyme — APPROVED** [review]

**Pull request merged**